src/thresholdtree.h

```diff
@@ -0,0 +1,195 @@
+        }
+        InnerIterator* ret = new InnerIterator();
+        all_inner_iterators.push_back(ret);
+        return ret;
+    }
+
+    // Return an InnerIterator object to the cache.
+    void ReleaseInnerIterator(InnerIterator* iter) {
+        available_inner_iterators.push_back(iter);
+    }
+
+    // Construct an InnerIterator object for a particular tree node.
+    InnerIterator* BuildInnerIterator(Node* node) {
+        InnerIterator *ret = AcquireInnerIterator();
+        ret->node = node;
+        if (node->IsLeaf()) {
+            accumulator.Push(node);
+            return ret;
+        }
+        int threshold = node->Threshold();
+        ret->children.resize(threshold);
+        for (int childnum = 0; childnum < threshold; childnum++) {
+            ret->children[childnum].position = childnum;
+            ret->children[childnum].inner_iterator = BuildInnerIterator(node->Child(childnum));
+        }
+        return ret;
+    }
+
+    // Move an InnerIterator to the next combinations it or its children allow.
+    // This will return false if there are no combinations left. In that case,
+    // InnerIterator will have been released.
+    bool IncrementInnerIterator(InnerIterator* iter) {
+        // First deal with the leaf case.
+        if (iter->node->IsLeaf()) {
+            accumulator.Pop(iter->node);
+            ReleaseInnerIterator(iter);
+            return false;
+        }
+        // Try to increment its child iterators first.
+        for (int childnum = iter->children.size() - 1; childnum >= 0; childnum--) {
+            if (IncrementInnerIterator(iter->children[childnum].inner_iterator)) {
+                for (int childnum2 = childnum + 1; childnum2 < iter->children.size(); childnum2++) {
+                    iter->children[childnum2].inner_iterator = BuildInnerIterator(iter->node->Child(iter->children[childnum2].position));
+                }
+                return true;
+            }
+        }
+        // If no child iterators can be incremented, quickly check whether we're done entirely.
+        if (iter->children[0].position + iter->children.size() == iter->node->Children()) {
+            ReleaseInnerIterator(iter);
+            return false;
+        }
+        // If not, move to the next combinations of children.
+        for (int childnum = iter->children.size() - 1; ; childnum--) {
+            if (iter->children[childnum].position + iter->children.size() != childnum + iter->node->Children()) {
+                iter->children[childnum].position++;
+                for (int childnum2 = childnum + 1; childnum2 < iter->children.size(); childnum2++) {
+                    iter->children[childnum2].position = iter->children[childnum2 - 1].position + 1;
+                }
+                for (int childnum2 = 0; childnum2 < iter->children.size(); childnum2++) {
+                    iter->children[childnum2].inner_iterator = BuildInnerIterator(iter->node->Child(iter->children[childnum2].position));
```

Comment:
Can you release the old `iter->children[childnum2].inner_iterator` before making a new one?